cpp/src/arrow/engine/substrait/extension_set.cc

```diff
@@ -204,6 +209,521 @@ const int* GetIndex(const KeyToIndex& key_to_index, const Key& key) {
+  auto cast_options = compute::CastOptions::Safe(std::move(output_type_desc.first));
+  return compute::call("cast", {substrait_convert_arguments(call)[0]}, std::move(cast_options));
+};
+
+ArrowToSubstrait arrow_cast_to_substrait = [] (const arrow::compute::Expression::Call& call, ExtensionSet* ext_set_) -> Result<substrait::Expression::ScalarFunction> {
+  substrait::Expression::ScalarFunction substrait_call;
+  ARROW_ASSIGN_OR_RAISE(auto function_reference, ext_set_->EncodeFunction("cast"));
+  substrait_call.set_function_reference(function_reference);
+  
+  auto arrow_to_type = std::dynamic_pointer_cast<compute::CastOptions>(call.options)->to_type;
+  ARROW_ASSIGN_OR_RAISE(auto substrait_to_type, ToProto(*arrow_to_type, false, ext_set_));
+  substrait_call.set_allocated_output_type(substrait_to_type.get());
+
+  auto expression = call.arguments[0];
+  ARROW_ASSIGN_OR_RAISE(auto value, ToProto(expression, ext_set_));
+  substrait_call.add_args()->CopyFrom(*value);
+  
+  return substrait_call;
+};
+
+// Datetime functions mapping
+SubstraitToArrow substrait_extract_to_arrow = [] (const substrait::Expression::ScalarFunction& call) -> Result<arrow::compute::Expression>  {
+  auto func_args = substrait_convert_arguments(call);
+  if(func_args[0].ToString() == "YEAR"){
+    return arrow::compute::call("year", {func_args[1]});
+  } else if (func_args[0].ToString() == "MONTH") {
+    return arrow::compute::call("month", {func_args[1]});
+  } else if (func_args[0].ToString() == "DAY") {
+    return arrow::compute::call("day", {func_args[1]});
+  } else {
+    return arrow::compute::call("second", {func_args[1]});
+  }
+};
+
+ArrowToSubstrait arrow_year_to_arrow = [] (const arrow::compute::Expression::Call& call, ExtensionSet* ext_set_) -> Result<substrait::Expression::ScalarFunction> {
+  substrait::Expression::ScalarFunction substrait_call;
+  ARROW_ASSIGN_OR_RAISE(auto function_reference, ext_set_->EncodeFunction("extract"));
+  substrait_call.set_function_reference(function_reference);
+  return arrow_convert_enum_arguments(call, substrait_call, ext_set_, "YEAR");
+};
+
+ArrowToSubstrait arrow_month_to_arrow = [] (const arrow::compute::Expression::Call& call, ExtensionSet* ext_set_) -> Result<substrait::Expression::ScalarFunction> {
+  substrait::Expression::ScalarFunction substrait_call;
+  ARROW_ASSIGN_OR_RAISE(auto function_reference, ext_set_->EncodeFunction("extract"));
+  substrait_call.set_function_reference(function_reference);
+  return arrow_convert_enum_arguments(call, substrait_call, ext_set_, "MONTH");
+};
+
+ArrowToSubstrait arrow_day_to_arrow = [] (const arrow::compute::Expression::Call& call, ExtensionSet* ext_set_) -> Result<substrait::Expression::ScalarFunction> {
+  substrait::Expression::ScalarFunction substrait_call;
+  ARROW_ASSIGN_OR_RAISE(auto function_reference, ext_set_->EncodeFunction("extract"));
+  substrait_call.set_function_reference(function_reference);
```

Comment:
All of these calls to `EncodeFunction` seem pretty repetitive.  Is there any way we can move this into the part that calls `GetArrowToSubstrait`?  Also, I don't see anything today that calls `GetArrowToSubstrait`